app/gui2/src/stores/graph/graphDatabase.ts

```diff
@@ -336,23 +333,51 @@ export class GraphDb {
   ) {
     const currentNodeIds = new Set<NodeId>()
     const body = [...functionAst_.bodyExpressions()]
-    body.forEach((outerAst, index) => {
-      const nodeId = nodeIdFromOuterExpr(outerAst)
-      if (!nodeId) return
-      const isLastInBlock = index === body.length - 1
+    const args = functionAst_.argumentDefinitions
+    const update = (
+      nodeId: NodeId,
+      ast: Ast.Ast,
+      isInput: boolean,
+      isOutput: boolean,
+      argIndex: number | undefined,
+    ) => {
       const oldNode = nonReactiveView(this.blockNodeLines.get(nodeId)?.data)
       if (oldNode) {
-        const node = resumeShallowReactivity<BlockLine>(oldNode)
-        if (oldNode.isLastInBlock !== isLastInBlock) node.isLastInBlock = isLastInBlock
-        if (oldNode.outerAst.id !== outerAst.id) node.outerAst = outerAst
+        const node = resumeShallowReactivity<NodeInfo>(oldNode)
+        if (oldNode.isOutput !== isOutput) node.isOutput = isOutput
+        if (oldNode.isInput !== isInput) node.isInput = isInput
+        if (oldNode.argIndex !== argIndex) node.argIndex = argIndex
+        if (oldNode.outerAst.id !== ast.id) node.outerAst = ast
       } else {
-        const data = shallowReactive({ isLastInBlock, outerAst })
+        const data = shallowReactive({ isOutput, outerAst: ast, isInput, argIndex })
         const stop = watchEffect(() =>
-          this.updateNodeStructure(nodeId, data.outerAst, data.isLastInBlock),
+          this.updateNodeStructure(
+            nodeId,
+            data.outerAst,
+            data.isOutput,
+            data.isInput,
+            data.argIndex,
```

Comment:
`updateNodeStructure` doesn't need to set `argIndex` if `updateNodes` does. The purpose of the separation is `updateNodeStructure` scans deep inside the node for properties derived from its internal AST, while `updateNodes` has to scan the whole function, but only shallowly, so it is only invalidated by high-level changes. In this case we have the `argIndex` data already without looking deeper into the AST.